norse/torch/functional/spikes_to_times_decoder.py

```diff
@@ -0,0 +1,98 @@
+import torch
+
+
+class SpikesToTimesDecoder(torch.nn.Module):
+    """
+    Module wrapper for spike to spike-time (or spike-index) decoder.
+
+    Decodes input spikes with shape [time_steps x batch_size x output_size] into spike indices
+    tensor with shape [spike_count x batch_size x output_size] in ascending order, e.g. for
+    spike_count = 1 returns only times of first spikes along each batch- and output-dimension.
+    If stated, converts indices along time axis into times with given dt value.
+    """
+
+    def __init__(
+        self,
+        spike_count: int = 1,
+        convert_indices_to_times: bool = True,
+        dt: float = 1e-3,
+    ):
+        """
+        Initialize decoder.
+        :param spike_count: number of elements (ascending) to return for each output channel.
+        :param convert_indices_to_times: Whether to return times or indices along input's time axis
+        :param dt: time step of input's time axis (needed for conversion of indices to times)
+        """
+        super().__init__()
+        self.spike_count = spike_count
+        self.convert_to_time = convert_indices_to_times
+        self.dt = dt
+        self.decoding_fn = self.ToSpikeTimes.apply
+
+    def forward(self, spike_input):
+        """
+        Call decoder custom autograd function
+
+        :param spike_input: spike tensor with 0s and 1s (indicating spikes)
+        """
+        spike_indices = self.decoding_fn(spike_input, self.spike_count)
+        if self.convert_to_time:
+            spike_indices = spike_indices * self.dt
+        return spike_indices
+
+    @staticmethod
+    class ToSpikeTimes(torch.autograd.Function):
+        """
+        Convert spike input with shape [time_steps x batch_size x output_size] to the indices of
+        spike times, shape [spike_count x batch_size x output_size]. If no spike is present, time
+        or index is set to inf.
+        """
+
+        @staticmethod
+        def forward(ctx, spike_input, spike_count):
+            """
+            Return indices of first spike_count spikes (if spike_count < spike_input.shape[0],
+            i.e. its time dim). If less than spike_count spikes happened along a output trace,
+            its remaining entries are set to inf.
+            """
+            indexed_spike_input = (
+                spike_input * torch.arange(1, spike_input.shape[0] + 1)[:, None, None]
+                - 1.0
+            )
+            indexed_spike_input[indexed_spike_input == -1.0] = torch.inf
+            spike_indices = torch.sort(indexed_spike_input, dim=0).values[
+                : spike_count
+                if spike_count < spike_input.shape[0]
+                else spike_input.shape[0]
+            ]
+            ctx.save_for_backward(spike_indices, spike_input)
+            ctx.shape = spike_input.shape
+            return spike_indices
+
+        @staticmethod
+        def backward(ctx, grad_output):
```

Comment:
Does this work with `torch.jit`?

Reply:
Actually, I don't have any experience with `torch.jit` yet. When using on the `ToSpikeTimes` autograd function directly `torch.jit.trace` fails. When implementing something like
```
def to_spike_times_fn(spike_input: torch.Tensor, spike_count: torch.Tensor):
    return ToSpikeTimes.apply(spike_input, spike_count)
```
in `norse/torch/functional/to_spike_times_decoder.py` (like in `norse/torch/functional/threshold.py` with `HeaviErfc` and `heavy_erfc_fn` for example)  and using `torch.jit.trace(to_spike_times_fn, example_inputs=...)` it works, but with the warning
```
TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  spike_indices = torch.sort(indexed_spike_input, dim=0).values[:spike_count if spike_count < spike_input.shape[0] else spike_input.shape[0]]
```

Using `torch.jit.script` gives some errors, I'm currently reading into this, but maybe you can give me a faster note on how `torch.jit` is normally used here, i.e. in norse or when working with norse functions/modules.

Reply:
Cool it works! I was hinting at creating a test for this, just so we know that this doesn't regress sometime in the future :-)

Reply:
Okay :) So do you think the extra function `to_spike_times_fn` should be added as mentioned above, or leave as is in the latest commit?

Reply:
I think we can leave that addition to a separate commit.

Reply:
TorchScript seems to not support jit compilation on custom autograd function directly atm, but using on `to_spike_times_fn` seems to work fine.
So I can add a test for the Module only, which should suffice since it uses `ToSpikeTimes.apply` i.e. the autograd function, ensuring jitting when used inside the Module works.

Reply:
The lack of support can be solved by adding the annotation `@torch.jit.ignore` (https://github.com/norse/norse/blob/master/norse/torch/functional/superspike.py#L21). 

Thanks for seeing this through btw! :-)